File: python/src/agent_api/types/responses.py
from __future__ import annotations

from typing import Any, Literal, TypedDict

from agent_api.types.common import AgentCapabilityPreference, ErrorInfo, ResponseStatus, Usage
from agent_api.types.tools import Tool


class ToolInvocationResult(TypedDict, total=False):
    id: str
    tool_call_id: str
    tool_name: str
    status: str
    error: ErrorInfo | None
    metadata: dict[str, Any]
    response_summary: str
    response_summary_mime_type: str


class AgentResponse(TypedDict, total=False):
    id: str
    object: Literal["response"]
    created_at: int
    completed_at: int | None
    status: ResponseStatus
    model: str
    output: list[dict[str, Any]]
    output_text: str
    usage: Usage
    error: ErrorInfo | None
    metadata: dict[str, Any]
    instructions: str | None
    tools: list[Tool]
    tool_choice: Any
    parallel_tool_calls: bool
    previous_response_id: str | None
    parent_response_id: str | None
    root_response_id: str | None
    prompt_cache_key: str | None
    store: bool
    background: bool
    tool_results: list[ToolInvocationResult]
    plan: Any


class ResponseListItem(TypedDict, total=False):
    id: str
    status: ResponseStatus
    created_at: int
    completed_at: int | None
    model: str
    preset: str
    input_preview: str
    root_response_id: str
    background: bool


class ResponseChildItem(TypedDict, total=False):
    id: str
    status: ResponseStatus
    created_at: int
    completed_at: int | None
    root_response_id: str
    model: str


class CancelResponse(TypedDict):
    interrupted: bool


class ModelCapabilities(TypedDict, total=False):
    provider: str
    supports_streaming: bool
    supports_tools: bool
    supports_json_schema: bool
    supports_reasoning: bool
    context_window: int
    pricing: dict[str, Any]
    metadata: dict[str, Any]


class Model(TypedDict, total=False):
    id: str
    object: Literal["model"]
    owned_by: str
    capabilities: ModelCapabilities


class PresetPolicy(TypedDict, total=False):
    plan_mode_preference: AgentCapabilityPreference
    sub_agent_preference: AgentCapabilityPreference
    allowed_tools: list[str]
    max_steps: int


class Preset(TypedDict, total=False):
    preset: str
    prompt_version: str
    preset_metadata: dict[str, Any]
    policy: PresetPolicy
    max_output_tokens: int
    default_model: str
    model_chain: list[str]


class PublicTool(TypedDict, total=False):
    object: Literal["tool"]
    name: str
    type: str
    description: str
    parameters: dict[str, Any]
    max_tokens: int
    max_tokens_per_page: int
    version: str

File: javascript/src/types/responses.ts
import type { AgentCapabilityPreference, ErrorInfo, ModelRoutingMode, ModelRoutingStrategy, OutputItemStatus, ResponseStatus, Usage } from "./common.js";
import type { ContentPart, Input, MemoryOptions, ReasoningConfig, ResponseFormat } from "./input.js";
import type { LocalSkillDescriptor, SkillReference } from "./skills.js";
import type { Tool } from "./tools.js";

export interface CallerContext {
  timezone?: string;
  locale?: string;
  locality?: string;
  extra?: unknown;
}

export interface SkillToolOptions {
  enabled?: boolean;
}

export interface ResponseCreateParamsBase {
  input: Input;
  instructions?: string;
  language_preference?: string;
  caller_context?: CallerContext;
  model?: string;
  models?: string[];
  model_routing?: ModelRoutingMode;
  routing_strategy?: ModelRoutingStrategy;
  preset?: "fast-search" | "pro-search" | "deep-research" | "advanced-deep-research" | string;
  max_output_tokens?: number;
  max_steps?: number;
  reasoning?: ReasoningConfig;
  response_format?: ResponseFormat;
  tools?: Tool[];
  tool_choice?: "auto" | "none" | "required" | Record<string, unknown>;
  parallel_tool_calls?: boolean;
  metadata?: Record<string, unknown>;
  store?: boolean;
  previous_response_id?: string;
  volume_id?: string;
  preferred_sites?: string[];
  skills?: SkillReference[];
  local_skills?: LocalSkillDescriptor[];
  skill_tool?: SkillToolOptions;
  prompt_cache_key?: string;
  memory?: MemoryOptions;
  plan_mode_preference?: AgentCapabilityPreference;
  sub_agent_preference?: AgentCapabilityPreference;
  safety_identifier?: string;
  user?: string;
  stream?: boolean;
}

export interface ResponseCreateParamsNonStreaming extends ResponseCreateParamsBase {
  stream?: false;
}

export interface ResponseCreateParamsStreaming extends ResponseCreateParamsBase {
  stream: true;
}

export type ResponseCreateParams = ResponseCreateParamsNonStreaming | ResponseCreateParamsStreaming;

export interface ToolInvocationResult {
  id?: string;
  tool_call_id?: string;
  tool_name?: string;
  status?: string;
  error?: ErrorInfo | null;
  metadata?: Record<string, unknown>;
  response_summary?: string;
  response_summary_mime_type?: string;
}

export interface MessageOutputItem {
  type: "message";
  id: string;
  status: OutputItemStatus;
  role: "assistant";
  content: ContentPart[];
}

export interface SearchResult {
  id: number;
  title: string;
  url: string;
  snippet: string;
  source?: string;
  date?: string;
  last_updated?: string;
}

export interface SearchResultsOutputItem {
  type: "search_results";
  queries?: string[];
  results: SearchResult[];
}

export interface URLContent {
  url: string;
  title: string;
  snippet: string;
}

export interface FetchURLResultsOutputItem {
  type: "fetch_url_results";
  contents: URLContent[];
}

export interface FunctionCallOutputItem {
  type: "function_call";
  id: string;
  status: OutputItemStatus;
  name: string;
  call_id: string;
  arguments: string;
  thought_signature?: string;
}

export type OutputItem =
  | MessageOutputItem
  | SearchResultsOutputItem
  | FetchURLResultsOutputItem
  | FunctionCallOutputItem;

export interface AgentResponse {
  id: string;
  object: "response";
  created_at: number;
  completed_at?: number | null;
  status: ResponseStatus;
  model: string;
  output: OutputItem[];
  output_text?: string;
  usage?: Usage;
  error?: ErrorInfo | null;
  metadata?: Record<string, unknown>;
  instructions?: string | null;
  tools?: Tool[];
  tool_choice?: unknown;
  parallel_tool_calls?: boolean;
  previous_response_id?: string | null;
  parent_response_id?: string | null;
  root_response_id?: string | null;
  prompt_cache_key?: string | null;
  store?: boolean;
  background?: boolean;
  tool_results?: ToolInvocationResult[];
  plan?: unknown;
  safety_identifier?: string;
}

export interface ResponseListItem {
  id: string;
  status: ResponseStatus;
  created_at: number;
  completed_at?: number | null;
  model?: string;
  preset?: string;
  input_preview?: string;
  root_response_id?: string;
  background?: boolean;
  safety_identifier?: string;
}

export interface ListResponsesParams {
  limit?: number;
  page_token?: string;
  safety_identifier?: string;
  user_id?: string;
}

export interface ListResponsesResponse {
  object: "list";
  data: ResponseListItem[];
  has_more: boolean;
  next_page_token?: string;
}

export interface ResponseChildItem {
  id: string;
  status: ResponseStatus;
  created_at: number;
  completed_at?: number | null;
  root_response_id?: string;
  model?: string;
}

export interface ListChildrenResponse {
  object: "list";
  data: ResponseChildItem[];
}

export interface ListEventsParams {
  after_sequence?: number;
  view?: "timeline" | "full";
}

export interface ListEventsResponse {
  data: import("./streaming.js").ResponseStreamEvent[];
}

export interface CancelResponse {
  interrupted: boolean;
}
